File: examples/javascript/hnsw_index_example.mjs
/** HNSW index -- hierarchical navigable small world graph with persistence. */

import { randomUUID } from "node:crypto";
import { MembrainIndex } from "membrain";

const DIMENSION = 128;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const config = { dimension: DIMENSION, m: 16, ef_construction: 200, ef_search: 100, distance_metric: "Cosine" };
const index = new MembrainIndex(DIMENSION, config);
let serialized;

try {
  for (let i = 0; i < 200; i++) index.add(randomUUID(), embed(i));
  console.log(`Indexed ${index.len()} vectors`);

  const results = index.search(embed(0.5), 5);
  for (const r of results) {
    console.log(`  id=${r.id.slice(0, 8)}... score=${r.score.toFixed(4)}`);
  }

  serialized = index.save();
  console.log(`\nSerialized: ${serialized.length} bytes`);
} finally {
  index.close();
}

const restored = MembrainIndex.load(serialized);
try {
  const results = restored.search(embed(0.5), 3);
  console.log(`Restored: ${restored.len()} vectors, top score=${results[0].score.toFixed(4)}`);
} finally {
  restored.close();
}

File: examples/python/hnsw_index_example.py
"""HNSW index -- hierarchical navigable small world graph with persistence."""

import random
import uuid

from membrain import MembrainIndex

DIMENSION = 128


def embed(seed: float) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(DIMENSION)]


def main() -> None:
    config = {"m": 16, "ef_construction": 200, "ef_search": 100, "distance_metric": "Cosine"}

    with MembrainIndex(dimension=DIMENSION, config=config) as index:
        for i in range(200):
            index.add(str(uuid.uuid4()), embed(seed=float(i)))

        print(f"Indexed {len(index)} vectors")

        results = index.search(embed(seed=0.5), k=5)
        for result in results:
            print(f"  id={result.id[:8]}... score={result.score:.4f}")

        serialized = index.save()
        print(f"\nSerialized: {len(serialized)} bytes")

    restored = MembrainIndex.load(serialized)
    results_after = restored.search(embed(seed=0.5), k=3)
    print(f"Restored: {len(restored)} vectors, top score={results_after[0].score:.4f}")
    restored.close()


if __name__ == "__main__":
    main()
